### PR TITLE
Feature/node fetching

### PR DESCRIPTION
**Pull request by tymnim**
Finished

### CODE REVIEW

File: dataProcessing.js [hunk 1]
```diff
@@ -1,42 +1,33 @@
...
-			return { description: step.text.slice(0, step.text.indexOf('.')), text: step.text }
-		})
+	 	return method.itemListElement.map(step => {
+	 		// desctiption is the first sentence
+	 		return { description: step.text.slice(0, step.text.indexOf('.')), text: step.text }
+	 	})
 	})
-	const ways = steps.map(st => ({ title, description: desc, steps: st, /* TODO: stepsToKeywords(st) */ }))
+	const ways = steps.map(st => ({ title, description: desc, steps: st, keywords: stepsToTagScores(st) }))
 	return ways
```
Comment (ANDREYDEN) [inline dataProcessing.js]:
```suggestion
	return steps.map(st => ({ title, description: desc, steps: st, keywords: stepsToTagScores(st) }))
```

File: dataProcessing.js [hunk 2]
```diff
@@ -1,42 +1,33 @@
...
+const { stepsToTagScores } = require('./tagExtraction.js')
+
+const htmlData = parse(data, { script: true })
+const result = extract(htmlData).map(res => JSON.stringify(res)).join(',')
+parentPort.postMessage(result)
 
 /*
  * extract function takes html and returns { title, description, steps } of the article
  * @html - reuslt 'node-html-parser.parse() includding script tags
  */
```
Comment (ANDREYDEN) [inline dataProcessing.js]:
```suggestion
/*
 * Takes html object and returns { title, description, steps } of the article
 * @param html {Object}
 * @returns {Object} - the resulting sections of the article 
 */
```

File: main.js [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
 // exits if no input or output file provided
-params.input || (function(){console.log('ERROR: missing input file\n to add use i: parameter.');process.exit(1)}())
+params.urls || (function(){console.log('ERROR: missing urls file\n to add use u: parameter.');process.exit(1)}())
 params.output || (function(){console.log('ERROR: missing output file\n to add use o: parameter.');process.exit(1)}())
 
+// TODO: params.log
+
 if (isMainThread) {
 	main(params)
 }
```
Comment (ANDREYDEN) [inline main.js]: Can we extract all this global stuff into a function?

File: main.js [hunk 4]
```diff
@@ -31,47 +36,70 @@ if (isMainThread) {
  * 	output - string, path to output file to write the result
  * }
  */
```
Comment (ANDREYDEN) [inline main.js]:
```suggestion
/*
 * @param urls {string} - the path to the file containing all the urls to fetch
 * @param output {string} - the path to the file that will contain the resulting objects
 */
```

File: main.js [hunk 5]
```diff
@@ -31,47 +36,70 @@ if (isMainThread) {
...
-				saveResultsToFile(data, output, () => console.log(`Finished: ${path}`))
-			}
-		})
-		// Error
-		worker.on('error', (err) => {
-			console.log('ERROR! ' + err)
+	IS_OUT_FILE_EMPTY = !(fs.existsSync(output) && fs.readFileSync(output, 'utf8').trim())
+	const urlList = fs.readFileSync(urls, 'utf8').split('\n').filter(p => p)
+	const minDelay = 3 // in seconds
+	const maxDelay = 5 // in seconds
```
Comment (ANDREYDEN) [inline main.js]: Can we use a single object for the delay?
```suggestion
	const delay = {
	        min: 3,
	        max: 5
	} // in seconds
```

File: main.js [hunk 6]
```diff
@@ -31,47 +36,70 @@ if (isMainThread) {
...
+
+	delayedFetch({ minDelay, maxDelay, urls: urlList, output })
+}
+
+function delayedFetch({ minDelay, maxDelay, urls, output, iterator = 0 }) {
```
Comment (ANDREYDEN) [inline main.js]:
```suggestion
/* Fetches the html of the url with a random delay and stores the result object into the output file
 * @param delay {Object} - the delay object
 * @param delay.min, delay.max - the minimu and maximum delays possible
 * @param urls {Array<string>} - array of urls to get the url from
 * @param iterator {int}- the index of the current url to fetch
 */
function delayedFetch({ delay, urls, output, iterator = 0 }) {
```

File: main.js [hunk 7]
```diff
@@ -31,47 +36,70 @@ if (isMainThread) {
...
+				setTimeout(() => delayedFetch({ minDelay, maxDelay, urls, output, iterator: iterator + 1 }), newDelay)
 			}
 		})
-	}
+	}).on('error', err => {
+		console.log("ERROR", err)
+	})
+}
+
+function processData({ data, url, output}) {
```
Comment (ANDREYDEN) [inline main.js]:
```suggestion
/* Creates a worker and saves result data to the output file
 * @param data {Object}
 * @param url {string} - the url currently being saved
 * @param output {string} - the path to the output file
 */
function processData({ data, url, output}) {
```

File: package.json [hunk 8]
```diff
@@ -20,7 +19,7 @@
     "wiki",
     "extractor"
   ],
-  "author": "Nahej Lefebvre and Tim Nimets",
+  "author": "Nahej Lefebvre, Tim Nimets and Andy Denysenko",
```
Comment (ANDREYDEN) [inline package.json]:
```suggestion
  "author": "Nahej Lefebvre, Tim Nimets and Andrii Denysenko",
```

File: main.js (hunk 3 above)
Comment (tymnim) [reply main.js]: Aren't we? Or you would make a function like extractParams which accepts process and returns param object?

**Comment by ANDREYDEN** [comment]
All good enough

**Pull request merged**